radicle-httpd/src/git.rs

```diff
@@ -0,0 +1,226 @@
+            let mut decoder = GzDecoder::new(&mut stdin);
+            let mut reader = body.reader();
+
+            io::copy(&mut reader, &mut decoder)?;
+            decoder.finish()?;
+        } else {
+            while body.has_remaining() {
+                let mut chunk = body.chunk();
+                let count = chunk.len();
+
+                io::copy(&mut chunk, &mut stdin)?;
+                body.advance(count);
+            }
+        }
+    }
+
+    match child.wait_with_output() {
+        Ok(output) if output.status.success() => {
+            tracing::info!("git-http-backend: exited successfully for {}", id);
+
+            let mut reader = std::io::Cursor::new(output.stdout);
+            let mut headers = HashMap::new();
+
+            // Parse headers returned by git so that we can use them in the client response.
+            for line in io::Read::by_ref(&mut reader).lines() {
+                let line = line?;
+
+                if line.is_empty() || line == "\r" {
+                    break;
+                }
+
+                let mut parts = line.splitn(2, ':');
+                let key = parts.next();
+                let value = parts.next();
+
+                if let (Some(key), Some(value)) = (key, value) {
+                    let value = &value[1..];
+
+                    headers
+                        .entry(key.to_string())
+                        .or_insert_with(Vec::new)
+                        .push(value.to_string());
+                } else {
+                    return Err(Error::Backend);
+                }
+            }
+
+            let status = {
+                tracing::debug!("git-http-backend: {:?}", &headers);
+
+                let line = headers.remove("Status").unwrap_or_default();
+                let line = line.into_iter().next().unwrap_or_default();
+                let mut parts = line.split(' ');
+
+                parts
+                    .next()
+                    .and_then(|p| p.parse().ok())
+                    .unwrap_or(StatusCode::OK)
+            };
+
+            let position = reader.position() as usize;
+            let body = reader.into_inner().split_off(position);
+
+            Ok((status, headers, body))
+        }
+        Ok(output) => {
+            tracing::error!("git-http-backend: exited with code {}", output.status);
+
+            if let Ok(output) = std::str::from_utf8(&output.stderr) {
+                tracing::error!("git-http-backend: stderr: {}", output.trim_end());
+            }
+            Err(Error::Backend)
+        }
+        Err(err) => {
+            panic!("failed to wait for git-http-backend: {err}");
+        }
+    }
+}
+
+#[cfg(test)]
+mod routes {
+    use std::net::SocketAddr;
+
+    use axum::body::Body;
+    use axum::extract::ConnectInfo;
+    use axum::http::Request;
+    use axum::http::{Method, StatusCode};
+    use tower::ServiceExt;
+
+    use crate::test;
+
+    #[tokio::test]
+    async fn test_issue_332() {
```

Comment:
```suggestion
    async fn test_invalid_route() {
```

Reply:
any benefits? I've seen this sort of usage in other projects such as `clippy`. `invalid_route` could mean many things, whereas `issue_322` gives you complete context.

Reply:
The benefit is knowing what test fails without having to look it up. `issue_322` gives you context only in context of github.

Reply:
[fixed](https://github.com/radicle-dev/heartwood/compare/06d5e0942902471569dfcd42fbc9a48453a14b77..22fb2d2afe976fca0ff88b13a66dd38b96b44e2c)